compiler/damlc/lib/DA/Cli/Damlc/Packaging.hs

```diff
@@ -79,22 +79,26 @@ createProjectPackageDb projectRoot (disableScenarioService -> opts) modulePrefix
       Logger.logDebug loggerH "package db is not up2date, reinitializing"
       clearPackageDb
 
+      stablePkgs <-
+        (>>= maybe (fail "Failed to generate stable packages") pure) $
+          withDamlIdeState opts loggerH diagnosticsLogger $ \ide ->
+            runActionSync ide $
+              runMaybeT $
+                useNoFileE GenerateStablePackages
+      let stablePkgIds :: Set LF.PackageId
+          stablePkgIds = Set.fromList $ map LF.dalfPackageId $ MS.elems stablePkgs
 
-      -- Register deps at the very beginning. This allows data-dependencies to
+      -- Register stable deps at the very beginning. This allows data-dependencies to
       -- depend on dependencies which is necessary so that we can reconstruct typeclass
       -- instances for a typeclass defined in a library.
-      -- It does mean that we can’t have a dependency from a dependency on a
-      -- data-dependency but that seems acceptable.
-      -- See https://github.com/digital-asset/daml/issues/4218 for more details.
-      -- TODO Enforce this with useful error messages
-      registerDepsInPkgDb depsDir dbPath
-
-      mbRes <- withDamlIdeState opts loggerH diagnosticsLogger $ \ide -> runActionSync ide $ runMaybeT $
-          (,) <$> useNoFileE GenerateStablePackages
-              <*> (fst <$> useE GeneratePackageMap projectRoot)
-      (stablePkgs, PackageMap dependenciesInPkgDb) <- maybe (fail "Failed to generate package info") pure mbRes
-      let stablePkgIds :: Set LF.PackageId
-          stablePkgIds = Set.fromList $ map LF.dalfPackageId $ MS.elems stablePkgs
+      registerStableDepsInPkgDb stablePkgIds depsDir dbPath
+      -- registerDepsInPkgDb depsDir dbPath
+
+      PackageMap dependenciesInPkgDb <-
```

Comment:
IIUC now this only contains the stable package dependencies, since those are the only ones we registered. I think this will break the calls to `decodeDalf` a bit further down, though it seems that's only used for collision and version checks, so maybe we can push that down to the point where we do have all dependencies registered

Reply:
I think you’re mixing up stable packages and dependencies. Stable packages are the ones with a fixed package id. those actually only exist in lf. In daml we expose them via daml-stdlib and daml-prim which are not stable LF packages.

Reply:
aha! got it, thank you